nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/state/server/ZooKeeperStateServer.java

```diff
@@ -198,6 +217,93 @@ public static ZooKeeperStateServer create(final NiFiProperties properties) throw
+
+        } else if (secureClientAddress != null && connectString == null) {
+            final int secureClientPort = secureClientAddress.getPort();
+            zkProperties.setProperty("secureClientPort", String.valueOf(secureClientPort));
+            logger.info("Secure client port set from ZooKeeper configuration, set port {}", secureClientPort);
+
+        } else if (secureClientAddress == null) {
+            final InetSocketAddress selectedServerAddress = getInetSocketAddress(connectString, "ZK not configured with secure port and NiFi ZK client connection string not usable.");
+            final int port = selectedServerAddress.getPort();
+            zkProperties.setProperty("secureClientPort", String.valueOf(port));
+            logger.info("Secure client port set from NiFi ZK connection string, set port {}", port);
+
+        } else {
+            final InetSocketAddress selectedServerAddress = getInetSocketAddress(connectString, "ZK configured with secure port but NiFi ZK client connection string not usable.");
+            if (secureClientAddress.getPort() != selectedServerAddress.getPort()) {
+                logger.warn("Potential mismatch between NiFi ZK client connection string and embedded ZK server secure port.");
+            } else {
+                logger.info("Matched ZK client connection string {} with embedded ZK server secure port: {}", connectString, secureClientAddress);
+            }
+        }
+
+        // If the ZK server connection factory isn't specified, set it to the one recommended for TLS:
+        final String cnxnSysKey = ServerCnxnFactory.ZOOKEEPER_SERVER_CNXN_FACTORY;
+        final String cnxnPropKey = "serverCnxnFactory";
+        if ((zkProperties.getProperty(cnxnPropKey) == null) && System.getProperty(cnxnSysKey) == null) {
+            zkProperties.setProperty(cnxnPropKey, SERVER_CNXN_FACTORY);
+        }
+
+        // Copy the NiFi properties if needed:
+        if (zkConfiguredPropCount == 0) {
+            zkPropKeys.forEach(zkKey -> {
+                final String nifiKey = ZOOKEEPER_TO_NIFI_PROPERTIES.get(zkKey);
+                final String logValue = (zkKey.endsWith(".password") ? "********" : niFiProperties.getProperty(nifiKey));
+                zkProperties.setProperty(zkKey, niFiProperties.getProperty(nifiKey));
+                logger.info("Mapped NiFi property '{}' to ZooKeeper property '{}' with value '{}'", nifiKey, zkKey, logValue);
+            });
+        } else {
+            logger.info("NiFi properties not mapped to ZooKeeper properties, all properties already set.");
+        }
+
+        // Recreate and reload the adjusted properties to ensure they're still valid for ZK:
+        peerConfig = new QuorumPeerConfig();
+        peerConfig.parseProperties(zkProperties);
+        return peerConfig;
+    }
+
+    private static InetSocketAddress getInetSocketAddress(String connectString, String message) throws ConfigException {
```

Comment:
As mentioned I don't think this is doing the right thing for clusters because of the `findFirst()`.